### DISCUSSION TITLE
Welcome to q-transformer Discussions!

### DISCUSSION THREAD

**Discussion by lucidrains** [lucidrains/q-transformer#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by lucidrains** [comment]
can anyone working in robotics comment on how significant this work is. I'll prioritize accordingly, thanks!

**Reply by ramkumarkoppu** [reply]
I am hoping to use Q-transformer for offline RL pretraing if my future project needs it

**Comment by justindujardin** [comment]
I don't work in robotics, but I'll definitely try this out with [Mathy](https://mathy.ai) when you think it's ready. I dropped the trained (A3C/AlphaZero) mathy agent examples a while ago in favor of swarm planning, but I'm interested in bringing them back to check out transformer-based agents that can work with large discrete action spaces.

**Reply by lucidrains** [reply]
oh hey Justin! i remember you from some repo long time ago

are you interested in trying it with Mathy due to the Q* rumors?

**Reply by lucidrains** [reply]
i haven't completely worked out how to apply it to language modeling yet, but it seems possible (maybe). should have a more definitive opinion once i complete this repository

**Reply by justindujardin** [reply]
Hey Phil, yeah we connected way back when I asked you to add support for extracting attention maps from your pytorch self-attention library or something similar (I was using it for Mathy at the time)

I came across this paper because of the Q* stuff, but I think it's genuinely interesting for Mathy. The curse of dimensionality was a problem since the action space is (Rules x Nodes) and it's sparse (most actions are invalid). DeepMind has done awesome stuff with dealing with large action spaces (AlphaStar comes to mind) so their approach here has me intrigued. Also, mathy naturally works with text (ascii math -> trees) so serializing its state for a transformer could be a nice fit.

I'm not at all sure this will line up, but it seems promising and I'm willing to try it out 😅

**Reply by lucidrains** [reply]
yea for sure, had my eyes on this paper for a while and finally getting around to it. no ideas if the Q* rumors are true or not, but at least the improvements for robotics seem ok (though not groundbreaking)

let me keep chipping away it! repository should be in a good place by week's end, and if ideas connect, i'll give it a shot for language models in another repo, which will be more in line with what you wish to do with Mathy

**Reply by justindujardin** [reply]
> let me keep chipping away it! repository should be in a good place by week's end

Awesome, I'll look at how it might connect with Mathy later this week when it's in a good place. 

> and if ideas connect, i'll give it a shot for language models in another repo, which will be more in line with what you wish to do with Mathy

If you introduce a language model variant later, I'll also check that out to see if it simplifies things for Mathy.

Do you know which version of `gym` you're going to use? I'm pretty sure it's had a turbulent couple of years recently and may have been superseded by the `gymnasium` package. Even within just the `gym` package there are breaking interface changes based on which version you use.

**Reply by lucidrains** [reply]
@justindujardin i don't think it can be tested on gym as i think all the games are single action, although maybe they could be reformulated to accept multiple actions for some given number of frames at a time? i'm not sure, will probably need to consult an RL researcher

**Reply by lucidrains** [reply]
btw, DuJardin is a cool name! i didn't even know "from the garden" can be a last name (took a little French in high school)

**Reply by justindujardin** [reply]
Ahh, I thought you were using gym because it's in the requirements and you mentioned testing on lunar lander which has a gym env: https://www.gymlibrary.dev/environments/box2d/lunar_lander/

**Reply by lucidrains** [reply]
@justindujardin yea, i'm figuring this out as i go along

lunar lander won't work, probably will need to do mujoco

**Reply by justindujardin** [reply]
Mujoco isn't bad since DeepMind made it free 👍 

_fyi: there are other envs in gym that have more complex action spaces, like [car racing](https://www.gymlibrary.dev/environments/box2d/car_racing/) and [bipedal walker](https://www.gymlibrary.dev/environments/box2d/bipedal_walker/)_

**Reply by lucidrains** [reply]
ohh nice, I'll check out bipedal walker!

**Reply by justindujardin** [reply]
@lucidrains How did your exploration turn out? Do you think this could be useful for Mathy? I added a [PPO example](https://mathy.ai/examples/mathy_ppo/) to demonstrate how I currently train models for Mathy environments. It would be excellent to finally find an architecture that can solve Mathy's "hard" poly simplification environments, but the action space appears too much for PPO (and friends) to handle. 😓

https://github.com/lucidrains/q-transformer/assets/101493/feee87d8-a52a-4bfd-9f9b-e30a1a884ef3

**Reply by lucidrains** [reply]
@justindujardin hey! yeah i'll get around to it over [here](https://github.com/lucidrains/llama-qrlhf). if you can understand the code there, maybe you can help pick up the slack and try it out?

truth is, i'm having a lot of trouble finding a knowledgeable community outside of the big communities like LLMs, vision, and audio somewhat. being met with crickets when it comes to RL, robotics, molecules, and of course, anything that touches bio / medicine. there is no open source advantage without a community and participation, and i'm not sure what to do about these niche fields, as much as i think they matter.

**Reply by lucidrains** [reply]
@justindujardin have you read the Alpha Geometry paper?

**Comment by justindujardin** [comment]
> @justindujardin hey! yeah i'll get around to it over [here](https://github.com/lucidrains/llama-qrlhf). if you can understand the code there, maybe you can help pick up the slack and try it out?

The trouble I have in contributing here is that I'm not sure how Mathy should integrate with LMs yet. Currently, it interfaces with models similar to any other Gym environment: it outputs observations in a floating-point format and only uses text as the input for the math parser that builds an expression tree.

For Mathy w/language models, I think perhaps interfacing with DSPy could be interesting, using LMs to handle the conversion of natural language inputs into ASCII-math expressions that Mathy can read and conversion of output states into natural language explanations based on the applied rule and affected expressions.

One other idea with Mathy is that you could use it to generate large datasets of step-by-step solutions to problems and use that dataset to fine-tune an LLM. Perhaps it overlaps with RLHF more in that kind of use case? I'm pretty ignorant here since I haven't spent much time with LLMs outside of daily ChatGPT use. 

> truth is, i'm having a lot of trouble finding a knowledgeable community outside of the big communities like LLMs, vision, and audio somewhat. being met with crickets when it comes to RL, robotics, molecules, and of course, anything that touches bio / medicine.

I feel this has been the case for a long time. I don't have much of an audience, but I've tried reaching out to RL people/companies for years now, and it seems like I'm the only one interested in math tutoring with RL 🤷 (or my limited ruleset makes the project appear too basic, or I didn't build a flashy web demo, or something else.)

> have you read the Alpha Geometry paper?

I saw it come by, but I still need to read it. I've been working with VAEs and Diffusion models lately, and my time for reading research papers is limited. Was it all hype or did it inspire hope? It's on my list for sure 📖

**Reply by lucidrains** [reply]
@justindujardin Alpha Geometry inspires a bit of hope, but not for the pure neural approach. They basically used a small transformer to cut down the search space for the main symbolic search. So a bit of a mixed bag of emotions. Yannic has a [great video](https://www.youtube.com/watch?v=ZNK4nfgNQpM) as usual

**Reply by lucidrains** [reply]
@justindujardin yea, one thing for sure, this idea of placing a cherry on top of the cake did not really transpire in the way LeCun imagined. and its hard to know whether the problem is with the cherry or the cake. i think its possibly both, but maybe it is too early to say

**Reply by lucidrains** [reply]
> The trouble I have in contributing here is that I'm not sure how Mathy should integrate with LMs yet. Currently, it interfaces with models similar to any other Gym environment: it outputs observations in a floating-point format and only uses text as the input for the math parser that builds an expression tree.

yea no worries, i'll try a few things once i circle back to this, maybe put up a simple arithmetic demo (if it works at all)
